platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ResolvedArtifactCollectingVisitor.java

```diff
@@ -60,10 +64,37 @@ public boolean requireArtifactFiles() {
     }
 
     public Set<ResolvedArtifactResult> getArtifacts() {
-        return artifacts;
+        return Sets.newLinkedHashSet(artifacts.values());
     }
 
     public Set<Throwable> getFailures() {
         return failures;
     }
+
+    private static class ArtifactIdentifier {
+        final ComponentArtifactIdentifier artifactIdentifier;
+        final File artifactFile;
+
+        public ArtifactIdentifier(ComponentArtifactIdentifier artifactIdentifier, File artifactFile) {
+            this.artifactIdentifier = artifactIdentifier;
+            this.artifactFile = artifactFile;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            ArtifactIdentifier that = (ArtifactIdentifier) o;
+            return Objects.equals(artifactIdentifier, that.artifactIdentifier) && Objects.equals(artifactFile, that.artifactFile);
```

Comment:
🤔 We should put an `@implSpec` on `ComponentArtifactIdentifier` stating that overriding `equals` in some meaningful way (not just identical instances) is now a requirement when implementing this interface.